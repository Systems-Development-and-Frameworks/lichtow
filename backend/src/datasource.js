import { DataSource } from "apollo-datasource";
import bcrypt, { hash } from "bcrypt";
import crypto from "crypto";

const saltRounds = 10;

export class Post {
    constructor(title, authorId) {
        this.id = crypto.randomBytes(16).toString("hex");
        this.voters = new Map();
        this.title = title;
        this.authorId = authorId;
    }
}
export class User {
    constructor(name, email, password) {
        this.id = crypto.randomBytes(16).toString("hex");
        this.name = name;
        this.email = email;
        this.password = password;
    }
    static async build(name, email, password) {
        const hashedPassword = await bcrypt.hash(password, saltRounds);
        return new User(name, email, hashedPassword);
    }
}

export class Neo4JDataSource extends DataSource {
    constructor() {
        super();
        this.posts = [];
        this.users = [];
    }

    initialize({ context }) {
        this.driver = context.driver;
    }

    allUsers() {
        return Promise.resolve(this.users);
    }

    async createUser(name, email, password) {
        const newUser = await User.build(name, email, password);
        this.users.push(newUser);
        return Promise.resolve(newUser);
    }

    getUser(userId) {
        return Promise.resolve(this.users.find((user) => user.id === userId));
    }

    getUserByEmail(email) {
        return Promise.resolve(this.users.find((user) => user.email === email));
    }

    allPosts() {
        return Promise.resolve(this.posts);
    }

    getPost(postId) {
        return Promise.resolve(this.posts.find((post) => post.id === postId));
    }

    createPost(title, authorId) {
        const newPost = new Post(title, authorId);
        this.posts.push(newPost);
        return Promise.resolve(newPost);
    }

    deletePost(postId) {
        const deletedPost = this.getPost(postId);
        this.posts = this.posts.filter((post) => post.id !== postId);
        return Promise.resolve(deletedPost);
    }

    upvotePost(postId, userId) {
        return this.getPost(postId).then((post) => {
            post.voters.set(userId, 1);
            return post;
        });
    }

    downvotePost(postId, userId) {
        return this.getPost(postId).then((post) => {
            post.voters.set(userId, -1);
            return post;
        });
    }
}
