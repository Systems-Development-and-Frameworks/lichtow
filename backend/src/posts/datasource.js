const { DataSource } = require("apollo-datasource");

class PostDataSource extends DataSource {
    constructor() {
        super();
        this.posts = [
            {
                id: 1,
                title: "Vue",
                votes: 0,
            },
            {
                id: 2,
                title: "Jurassic Park",
                votes: 2,
            },
        ];
    }

    initialize({ context }) {}

    allPosts() {
        return this.posts;
    }
    createPost(data) {}
    upvotePost(id, user) {}
}
module.exports = {
    PostDataSource
}
