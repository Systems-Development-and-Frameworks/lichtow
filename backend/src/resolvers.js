import { UserInputError } from "apollo-server";
import bcrypt from "bcrypt";

const resolvers = {
    Query: {
        users: (parent, args, context) => context.dataSources.db.allUsers(),
        posts: (parent, args, context) => context.dataSources.db.allPosts(),
    },
    Mutation: {
        signup: async (_, args, { dataSources }) => {
            const name = args.name;
            const email = args.email;
            const password = args.password;
            if (password.length < 9) {
                throw new UserInputError("Password is too short", { invalidArgs: password });
            }
            const users = await dataSources.db.allUsers();
            if (users.find((user) => user.email === email)) {
                throw new UserInputError("User with this email already exists", { invalidArgs: email });
            }
            const newUser = await dataSources.db.createUser(name, email, password);
            return newUser.id;
        },
        login: async (_, args, { dataSources, jwt }) => {
            const email = args.email;
            const password = args.password;
            const user = await dataSources.db.getUserByEmail(email);
            if (!user) {
                throw new UserInputError("No user with this email", { invalidArgs: email });
            }
            const isCorrectPassword = await bcrypt.compare(password, user.password);
            if (!isCorrectPassword) {
                throw new UserInputError("Password is incorrect");
            }
            let token = jwt.sign({ userId: user.id }, process.env.JWT_SECRET);
            return token;
        },
        write: async (_, args, { dataSources, userId }) => {
            const title = args.post.title;
            const user = await dataSources.db.getUser(userId);
            if (!user) {
                throw new UserInputError("Invalid user", { invalidArgs: userId });
            }
            return await dataSources.db.createPost(title, userId);
        },
        upvote: async (_, args, { dataSources, userId }) => {
            const postId = args.id;
            const user = await dataSources.db.getUser(userId);
            if (!user) {
                throw new UserInputError("Invalid user", { invalidArgs: userId });
            }
            const post = await dataSources.db.getPost(postId);
            if (!post) {
                throw new UserInputError("Invalid post", { invalidArgs: postId });
            }
            return await dataSources.db.upvotePost(postId, userId);
        },

        // delete: async (parent, args, context) => {
        //     const id = args.id;
        //     const post = await context.dataSources.db.getPost(id);
        //     if (!post) {
        //         throw new UserInputError("Invalid post", { invalidArgs: id });
        //     }
        //     return await context.dataSources.db.deletePost(id);
        // },

        // downvote: async (parent, args, context) => {
        //     const name = args.voter.name;
        //     const id = args.id;
        //     const downvoter = await context.dataSources.db.getUser(name);
        //     if (!downvoter) {
        //         throw new UserInputError("Invalid user", { invalidArgs: name });
        //     }
        //     const post = await context.dataSources.db.getPost(id);
        //     if (!post) {
        //         throw new UserInputError("Invalid post", { invalidArgs: id });
        //     }
        //     return await context.dataSources.db.downvotePost(id, name);
        // },
    },
    Post: {
        author: async (obj, args, context) => {
            return await context.dataSources.db.getUser(obj.authorId);
        },
        votes: async (obj, args, context) => {
            let values = Array.from(obj.voters.values());
            let votes = values.reduce((sum, number) => sum + number, 0);
            return votes;
        },
    },
    User: {
        posts: async (obj, args, context) => {
            const allPosts = await context.dataSources.db.allPosts();
            return allPosts.filter((post) => post.authorId === obj.name);
        },
    },
};

export default resolvers;
