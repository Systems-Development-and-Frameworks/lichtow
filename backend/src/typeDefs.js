import { gql } from "apollo-server";

const typeDefs = gql`
    type Post {
        id: ID!
        title: String!
        votes: Int!
        author: User!
    }

    type User {
        id: ID!
        name: String!
        email: String!
        posts: [Post]
    }

    type Query {
        posts: [Post]
        users: [User]
    }

    type Mutation {
        write(post: PostInput!): Post
        delete(id: ID!): Post
        upvote(id: ID!): Post
        downvote(id: ID!): Post

        login(email: String!, password: String!): String
        signup(name: String!, email: String!, password: String!): String
    }

    input PostInput {
        title: String!
    }
`;

export default typeDefs;
