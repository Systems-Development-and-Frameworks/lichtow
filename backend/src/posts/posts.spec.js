import  { ApolloServer, gql } from "apollo-server";
import  { createTestClient } from "apollo-server-testing";
import  typeDefs  from "../typeDefs";
import  resolvers  from "../resolvers";
import  { PostDataSource } from "./datasource";

const dataSources = {
    postDatasource: new PostDataSource(),
};


const GET_POSTS = gql`
    {
        posts {
            id
            votes
            title
        }
    }
`;

it("fetches posts", async () => {
    const server = new ApolloServer({
        typeDefs: [typeDefs],
        resolvers: [resolvers],
        dataSources: () => {
            return dataSources;
        },
    });

    const { query } = createTestClient(server);

    const res = await query({ query: GET_POSTS });
    expect(res).toMatchSnapshot();
});
