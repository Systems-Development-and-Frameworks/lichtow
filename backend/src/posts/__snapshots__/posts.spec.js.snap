// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches posts 1`] = `
Object {
  "data": Object {
    "posts": Array [
      Object {
        "id": "1",
        "title": "Vue",
        "votes": 0,
      },
      Object {
        "id": "2",
        "title": "Jurassic Park",
        "votes": 2,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
